lib/Wasmtime.hs

```diff
@@ -1380,64 +1406,92 @@ newFunc ::
     MonadPrim s m,
     PrimBase m
   ) =>
-  Context s ->
+  Store s ->
   -- | 'Funcable' Haskell function.
   f ->
   m (Func s)
-newFunc ctx f = unsafeIOToPrim $ withContext ctx $ \ctx_ptr ->
-  withFuncType funcType $ \functype_ptr -> do
-    -- TODO: We probably need to extend the Store with a mutable set of FunPtrs
-    -- that need to be freed using freeHaskellFunPtr when the Store is collected!!!
-    callback_funptr <- mk'wasmtime_func_callback_t callback
-    alloca $ \(func_ptr :: Ptr C'wasmtime_func_t) -> do
-      c'wasmtime_func_new ctx_ptr functype_ptr callback_funptr nullPtr nullFunPtr func_ptr
-      MkFunc <$> peek func_ptr
+newFunc store f =
+  unsafeIOToPrim $
+    withStore store $ \ctx_ptr ->
+      withFuncType funcType $ \functype_ptr -> do
+        callback_funptr :: FunPtr FuncCallback <-
+          newFuncCallbackFunPtr (storeFinalizeRef store) callback
+        func <- MkFunc <$> mallocForeignPtr
```

Comment:
I am still not sure where this is free'd. the lines above concern the callback_funptr, for which I can see the frees. but not for this one.